src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordPrintMembers.cs

```diff
@@ -0,0 +1,277 @@
+                    }
+                }
+
+                block.Add(F.Return(F.Literal(true)));
+
+                F.CloseMethod(F.Block(block.ToImmutableAndFree()));
+            }
+            catch (SyntheticBoundNodeFactory.MissingPredefinedMember ex)
+            {
+                diagnostics.Add(ex.Diagnostic);
+                F.CloseMethod(F.ThrowNull());
+            }
+        }
+
+        internal static MethodSymbol? FindValidPrintMethod(TypeSymbol containingType, CSharpCompilation compilation)
+        {
+            if (containingType.IsErrorType() || containingType.IsObjectType())
+            {
+                return null;
+            }
+
+            MethodSymbol? candidate = null;
+            var accessibility = (containingType.BaseTypeNoUseSiteDiagnostics.IsObjectType() && containingType.IsSealed) ?
+                Accessibility.Private :
+                Accessibility.Protected;
+
+            var stringBuilder = TypeWithAnnotations.Create(compilation.GetWellKnownType(WellKnownType.System_Text_StringBuilder));
+
+            foreach (var member in containingType.GetMembers(WellKnownMemberNames.PrintMembersMethodName))
+            {
+                if (member is MethodSymbol { DeclaredAccessibility: var foundAccessibility, IsStatic: false, ParameterCount: 1, Arity: 0, ParameterTypesWithAnnotations: var parameters } method &&
+                    foundAccessibility == accessibility &&
+                    parameters[0].Equals(stringBuilder, TypeCompareKind.AllIgnoreOptions))
+                {
+                    if (candidate is object)
+                    {
+                        // An ambiguity case, can come from metadata, treat as an error for simplicity.
+                        return null;
+                    }
+
+                    candidate = method;
+                }
+            }
+
+            if (candidate is null ||
+                !(containingType.IsSealed || candidate.IsOverride || candidate.IsVirtual) ||
+                candidate.ReturnType.SpecialType != SpecialType.System_Boolean)
+            {
+                return null;
+            }
+
+            return candidate;
+        }
+
+        internal static void VerifyOverridesPrintMembersFromBase(MethodSymbol overriding, DiagnosticBag diagnostics)
+        {
+            if (overriding.ContainingType.BaseTypeNoUseSiteDiagnostics.IsObjectType())
+            {
+                return;
+            }
+
+            bool reportAnError = false;
+
+            if (!overriding.IsOverride)
+            {
+                reportAnError = true;
+            }
+            else
+            {
+                var overridden = overriding.OverriddenMethod;
+
+                if (overridden is object &&
+                    !overridden.ContainingType.Equals(overriding.ContainingType.BaseTypeNoUseSiteDiagnostics, TypeCompareKind.AllIgnoreOptions))
```

Comment:
is this check just to catch a situation where an intermediate base type doesn't override PrintMembers?

Reply:
Yes. I'll be adding a test shortly (realized I missed it), `ToString_BadBase_PrintMembersNotOverriddenInBase`

---
In reply to: [469415941](https://github.com/dotnet/roslyn/pull/46616#discussion_r469415941) [](ancestors = 469415941)